### PR TITLE
Normalizes the handling of steps and missing tools for the workflow editor

### PR DESCRIPTION
**Pull request by guerler**
This PR unifies the handling and storage of workflow modules and their states. Now, preruntime and runtime states are aligned with the front-end form rendering. The overall goal of this PR is to streamline the module handling process.

Please see also comments below within this thread. 

Here is a list of changes so far:

1. All modules are handled through a single base class which uniformly encodes and decodes preruntime and runtime states i.e. the SimpleWorkflowModule layer for non-tool modules has been removed.

2.  Preruntime states for non-tool modules specify the same input parameter types as tool modules i.e. we could provide xml or yaml descriptions for non-tool module input parameters similar to tool.xmls.

3. Tool and node input errors are not stored or exported anymore. We evaluate them when module states are injected.

4. Datalists which provide suggestions for collection input types are handled through the same process as regular tool input parameters i.e. the TextToolParameter has been augmented. The front-end implementation for displaying these options is not part of this PR but will be implemented in a follow up.

5. All workflow modules return a form object instead of html. The form object contains input definitions which are rendered through the regular js-form view. The editor form mako for non-tool makos has been removed. We also removed the corresponding front-end view for non-tool modules EditorFormView.

6. If no workflow node is selected in the editor view, the global workflow options form is shown instead of the no node selected notification message.

7. Missing tools are handled like regular tools without the option to edit parameters or to save the workflow.

8. Several functions have been removed from the module handling process as a result of the unification  such as the separate get_new_module_info (uses build_module instead), editor_form_post (uses build_module instead), new (uses from_dict instead), update_state (uses recover_state instead), get_runtime_input_dicts, _abstract_config_form, default_state callers.

9. For backward compatibility, modules states are loaded using safe_loads instead of loads. Once stored, workflow module states are upgraded to the unified state format.

ping @jmchilton

### CODE REVIEW

**Comment by martenson** [comment]
xref https://github.com/galaxyproject/galaxy/pull/3064 https://github.com/galaxyproject/galaxy/pull/2823

**Comment by guerler** [comment]
Thanks @martenson

**Comment by guerler** [comment]
@galaxybot test this

**Comment by guerler** [comment]
@galaxybot test this

**Comment by martenson** [comment]
@galaxybot test this

**Comment by guerler** [comment]
@galaxybot test this

**Comment by guerler** [comment]
@galaxybot test this

**Comment by guerler** [comment]
@galaxybot test this

**Comment by guerler** [comment]
@galaxybot test this

**Comment by bgruening** [comment]
There is currently no indication which tools have been updated, maybe we can add a list of those tools to this text?
```
Some tools are being executed with different versions compared to those available when this workflow was last saved because the other versions are not or no longer available on this Galaxy instance. To upgrade your workflow and dismiss this message simply edit the workflow and re-save it.
```

**Comment by guerler** [comment]
Yeah, that would be good. Your comment is related to the run worflow form right?

**Comment by bgruening** [comment]
I haven't realized this yet, but with this version you don't need to have an 'input-dataset-box' in your workflow. The first tool can simply use the one from the history! This is great! But I wonder if this is intended :)

@guerler yes related to the worflow form.

**Comment by guerler** [comment]
Yes it is. We should open a separate issue to extend the run workflow form notification message. I can work on it tho.

File: lib/galaxy/managers/workflows.py [hunk 1]
```diff
@@ -842,35 +819,7 @@ def __walk_step_dicts( self, data ):
 
             yield step_dict
 
-    def __track_module_from_dict( self, trans, steps, steps_by_external_id, step_dict, exact_tools=False ):
```
Comment (jmchilton) [inline lib/galaxy/managers/workflows.py:845]: I dislike this refactoring. I get the function has a crappy name - but the operations should be distinct. I prefer the simpler lower mccabe complexity functions - and the ability to see the relatively stateless __module_from_dict from the larger function that tracks a bunch of state as it goes.

This is a like -0 thing, I'm not going to hold up the PR based on it - just expressing my desire and design aesthetic.

**Comment by jmchilton** [comment]
Thanks for the contribution. The features you added are important and I like being able to deal with missing tools - that is important. But the refactoring aspects frustrate me. **It seems like you've normalized the modules by making it all as messy as the tool module.** Take the name method for instance, or how state is handled, or how inputs are treated - these things used to be cleaner IMO and now they are more complex. We hid the complexity of tools in the tool modules and the others were simpler. You are now constructing XML in a Python string and parsing that in order to build a representation of a form on the backend to transmit to the front end. This is backwards - we should move toward defining these forms on the front end, not toward making it more dependent on our XML. The previous representation of the form was not great because it was on the backend, but at least it was cleaner.

There are serious regressions in the way non-tool modules are displayed. They used to have icons in the editor view on the side and title was editable and had logic for managing the labels. All module types had that except tools - because I could manage the complexity of adding that for those modules but the tool form was intractable for me. This goes to something you said before:

> The overall goal of this PR is to streamline the module handling process.

I feel like you've legitimately made it easier for @guerler to modify/extend modules - but you've made it harder for the rest of us - or at least me. Progress on things like extending the input module to contain booleans for "optional" and "multiple" or implementing a GUI for the "InputParameter" modules halted because I didn't know how to modify editor_generic.mako - after this PR I'm going to have even more trouble I think. (I may be wrong, if I just stop fighting it and think of these things in terms of a tool form - maybe I'll be fine.)

I'm a -1 until the module type icons and label management are added back to the workflow editor and there is some mechanism in place to prevent data list stuff coming from tools accidentally (either we flag inputs defined from workflow module as having more options - or we add these entries to the tool schema and add a test tool demonstrating them).

Adding the label management stuff to this PR though will have the added advantage of allowing tool steps to define labels in the GUI - this is part of #1325. (I'm aware I may be proven wrong and making everything like tool modules might be the right direction - it will certainly have some upshots such as resolving this label problem.)

At any rate - after those two things are done - I'll be a -0 on this PR - but I will click the merge button for you if you'd like. I appreciate the effort and I have faith you'll continue to build out the GUI for these modules to catch up with backend features available.

**Comment by jmchilton** [comment]
Tool authors (e.g. @bgruening @nsoranzo @erasche @blankenberg) this PR introduces an implicit ability to allow options inside of a ``text`` type. Should we incur the complexity of adding that to the public tool definition language or should we restrict it to be just an internal thing? I'm guessing the bar I would like to set for exposing it is - does anyone have a target tool that this would improve and will the add it to that tool once this is in Galaxy?

I'm happy to take over that aspect of this PR and implement either adding it to the language or restricting it to be internal use only.

```
            <param name="collection_type" label="Collection type" type="text" value="list">
                <option value="list">List of Datasets</option>
                <option value="paired">Dataset Pair</option>
                <option value="list:paired">List of Dataset Pairs</option>
            </param>
```

**Comment by nsoranzo** [comment]
@jmchilton That would be like a suggestions list?

**Comment by jmchilton** [comment]
@nsoranzo Yup - like used for collection types in the workflow editor. I guess the idea would over time to come up with a better, more portable representation of the lists though.

**Comment by bgruening** [comment]
A made up use-cases would be a text box to specify amino acids, a tool author would probably include 20 in the list, but they are way more that a user could enter.

I will think more about this.

**Comment by guerler** [comment]
@jmchilton Thank you for your excellent review. I am aware of the ui regressions and would work instantly after this PR on it if thats ok with you. It should be straightforward and ready before the next release and this PR is already fairly complex. Regarding handling all module states as tool states, it is something I thought quite a bit about. We have four states of which only one, the non-tool module preruntime state is handled differently and that produces significant overhead and layers. Non-tool module runtime states are already handled as tool states and I believe that regular tool states are more powerful on the long run which is why I aligned this 1-out-of-4 handling mechanism.

**Comment by hexylena** [comment]
A suggestion box but allowing for other inputs? I would use that

- jbrowse tool allows choosing from GFF3 fields, which should be displayed as a feature label (e.g. can display ID field or Name field).  Right now it's a plain text box, would be better with suggestions.
- jbrowse allows for selecting an icon to represent certain menus, the list is gigantic but showing just a handful of popular ones would be nice, and allowing people to use the full list only by browsing the documentation.

**Comment by jmchilton** [comment]
@guerler You could put it that way or you could say 5 modules types had simple forms for configuration before this PR and now all 6 module types are defining complex tool parameters and state for configuration.

I really do not want this PR to be merged with that regression. How about we make a deal? I will rescind the -1 however if you promise that it is fixed before "the next release" and if it isn't you do the work of rolling it back this PR before "the next release" is announced.

@erasche Fair enough - that is all I need. I'll add an example of this to the example tools and update the language in the XSD via a PR to this branch.

**Comment by guerler** [comment]
Yeah, I should need just 2-3 days to fix those two ui regressions. So this week it should be good to go.

**Comment by jmchilton** [comment]
@guerler I'm doing some testing on the new tool stuff and I realized actually the drop down for the collection type doesn't appear anymore either - is that known?

**Comment by guerler** [comment]
Yes. I want to implement something which also works for Safari, also before the release. If I dont find a solution which also works for Safari, I will just use regular HTML5 datalists again. Its mentioned in 4. in the above list of changes.

**Comment by jmchilton** [comment]
@guerler I opened a PR for what I want to see in the tool stuff and to tweak a logging level https://github.com/guerler/galaxy/pull/17.

**Comment by guerler** [comment]
Thanks @jmchilton

**Comment by gregvonkuster** [comment]
Just wondering: if this PR is merged, will anything be broken on the dev branch?  I'm wondering due to the regression fixes, missing features like select lists, etc that will be done before "the next release" is announced, which I assume is a ways out.  I have several systems that I'm running from the dev branch, but I'll hold off on updating them if this PR will pose problems for workflows, etc.  ;)

**Comment by guerler** [comment]
It should not cause any issues but it would be great if you can run tests. The regressions are very minor ui regressions related to displaying icons, editing the label in the title of the form instead of the text field used in this PR and HTML5 datalists which are not supported in Safari and only used to display suggestions when editing the collection type text field.

**Comment by gregvonkuster** [comment]
Ok, thanks for the info @guerler ;)

**Comment by guerler** [comment]
@jmchilton we can also merge this after branching 17.01 if you think thats better.

**Comment by jmchilton** [comment]
@guerler I think it would be better - I will try to do what I can to ensure we branch 17.01 as soon as possible.

**Comment by guerler** [comment]
Sounds good. Thanks a lot.

**Comment by jmchilton** [comment]
Switched milestone in Github to reflect this.

**Comment by jmchilton** [comment]
Merged with reservation previously discussed but very thankful to @guerler for waiting so long on this.

**Pull request merged**